### DISCUSSION TITLE
Locking to a particular PHP version

### DISCUSSION THREAD

**Discussion by aquavark** [oerdnj/deb.sury.org#1888] [Q&A]
Hi,

I obviously asked this in the wrong place originally, in issues. Yes I read the FAQ, I couldn't see that it fitted in bug or feature, but I did somehow miss the template step when I opened a new issue, not sure how. Sorry for that.

As in the issue, I am appreciative of the time and effort that goes into maintaining this repo and what is much needed functionality to keep PHP up to date (given that Bookworm still seems a way off and PHP 7.4 is EOL now).

I am looking to see if anyone has a good suggestion as to locking the PHP version coming from this repo to PHP 8.1, as whilst I'm testing 8.2 on my development systems I have a number of servers running Bullseye but with 8.1 on, and now if they try to update they get 8.2 as an upgrade which then caused some applications to fail (in part because not all of the modules were available to update or they need rebuilding for 8.2 e.g. the mssql libraries from Microsoft - which isn't done with the routine update). I don't want to drop the repo as I don't want to lose any security updates. I'm hoping to have completed the testing and get onto 8.2 but that won't happen for a few months (holiday season followed by change control needs). I'm not the one running the server updates, and those that do are not trained to spot the subtleties of keeping PHP at a given version.

I have tried to use pinning as best as I could, but I'm still seeing a number of things attempting to update, in particular if I've installed any of the php-(library names) which defaults to the current version, and also when installing the individual versions it appeared to pull in php-common, which then pulled php8.2 again. Looking to see if I've missed something obvious (like a reference to a specific version repo).

Thanks in advance for any suggestions/help.

**Comment by oerdnj** [comment]
These doesn't answer your questions?

* https://github.com/oerdnj/deb.sury.org/wiki/Managing-Multiple-Versions
* https://github.com/oerdnj/deb.sury.org/wiki/Frequently-Asked-Questions#why-is-phpdefaultversion-cli-always-installed

You don't need to lock anything down - the PHP 8.2 packages will only get installed if you have the META packages (e.g. `php-fpm`, or `php-cli` installed).

And even then - those will get bumped only if you run `apt-get dist-upgrade`. If you don't want new packages to get installed use `apt-get upgrade` (this has been always the case):
```
       upgrade
           upgrade is used to install the newest versions of all packages currently installed on the system from the sources enumerated in /etc/apt/sources.list. Packages currently installed with new versions available are
           retrieved and upgraded; under no circumstances are currently installed packages removed, or packages not already installed retrieved and installed. New versions of currently installed packages that cannot be
           upgraded without changing the install status of another package will be left at their current version. An update must be performed first so that apt-get knows that new versions of packages are available.
```

**Comment by aquavark** [comment]
Thanks for the reply. 

No - they didn't. I read them before asking.

I'll recheck for the meta packages and manually install all the individual packages as 8.1 only as the next step.

**Comment by aquavark** [comment]
Thanks again - there were some lurking metapackages that were causing this issue, removing and reinstalling as the specific PHP version has solved this.

**Reply by oerdnj** [reply]
Perhaps, you can offer a text that I can put into the FAQ?

**Reply by aquavark** [reply]
Installing and maintaining a specific version release:

If you require a specific PHP version and want it kept up to date, but not just receive the current release that gets updated when PHP new releases occur (for example the release of PHP 8.2 occurs when you want to stay on PHP 8.1) it is required that when you install PHP and any modules you only install version specific packages and not the generic packages. For example, most packages starting php- (e.g. php itself, and php-mbstring) are the generic and will always default to the latest release, whereas selecting php8.1-clim php8.1-fpm, php8.1-mbstring will install (and update) the specific 8.1 release (and in the case of PHP itself, the specific CLI, FPM etc version of same - note the entry [in this FAQ] concerning why some packages will pull in others automatically).

Using this method will ensure that critical installs, e.g. to servers, are not suddenly replaced with a new version of PHP that may contain breaking changes or cause unexpected failures (e.g. a package not being available for your new release or an independent package that needs recompiling e.g. through PECL).

Also, this means that you do not need to try pinning PHP, and that you are well set up to support multiple PHP versions on your system (e.g. to allow website installs to point to specific php-fpm versions).

**Reply by aquavark** [reply]
Swapping from metapackages to version specific:

Use your package manager to search for php-* packages - you will need to remove these for any you have installed outside of the PHP base. This includes for PHP itself - install the version specific (apt install php8.1-cli, apt-install php8.1-fpm, don't forget to install the apache module if you need it also version specific)

For each library that you identify, use your package manager to remove it (but don't purge - you want to keep your settings), and immediately add back the version specific library, e.g.

(Debian)
apt remove php-mbstring
apt install php8.1-mbstring

Another way to check if you have libraries remaining that need to be version fixed is to do an update and upgrade (but don't actually permit it to install yet) or list upgraded packages and see which are going to be version pushed to a later release (i.e. a 7.4 or 8.1 package now being pushed to an 8.2 - that will be from the meta package install). 

Caveat: php-common appears to be installed and updated no matter what and has not been seen to be an issue.

To double check that you've covered everything, do a php[version] -m to list the modules installed before you start and removal, and compare with a php[version] -m list when you've finished - e.g. php8.1 -m (this is also a good technique to compare during updgrading, e.g. php7.4 -m compared to php8.1 -m - make sure you've got everything installed for the new release).

**Comment by oerdnj** [comment]
Thanks, I've added your texts to the wiki with just few style adjustments.
